cl/phase1/stages/clstages.go

```diff
@@ -518,99 +513,67 @@ func ConsensusClStages(ctx context.Context,
 					)
 					respCh := make(chan *peers.PeeredObject[[]*cltypes.SignedBeaconBlock], 1024)
 					errCh := make(chan error)
-					sources := []persistence.BlockSource{cfg.gossipSource, rpcSource}
 
-					// if we are more than one block behind, we request the rpc source as well
-					if totalRequest > 2 {
-						sources = append(sources, rpcSource)
-					}
 					// 15 seconds is a good timeout for this
 					ctx, cn := context.WithTimeout(ctx, 25*time.Second)
 					defer cn()
 
-					// we go ask all the sources and see who gets back to us first. whoever does is the winner!!
-					for _, v := range sources {
-						sourceFunc := v.GetRange
-						go func(source persistence.BlockSource) {
-							if _, ok := source.(*persistence.BeaconRpcSource); ok {
+					go func() {
+						select {
+						case <-time.After((time.Duration(cfg.beaconCfg.SecondsPerSlot) * time.Second) / 2):
+						case <-ctx.Done():
+							return
+						}
 
+						for {
+							var blocks *peers.PeeredObject[[]*cltypes.SignedBeaconBlock]
```

Comment:
think I need to read this part to get more understanding what you did ..

Reply:
the FSM is interesting but a bit messy

Reply:
clstages is cursed lol

Reply:
but overall i feel it's a good design here to apply state machine. Do you think caplin is actually not a state transitioning pattern?

Reply:
it is, it is just a bit convuluted